presto-analyzer/src/main/java/com/facebook/presto/sql/analyzer/BuiltInQueryPreparer.java

```diff
@@ -0,0 +1,167 @@
+import com.facebook.presto.sql.parser.ParsingException;
+import com.facebook.presto.sql.parser.SqlParser;
+import com.facebook.presto.sql.tree.Execute;
+import com.facebook.presto.sql.tree.Explain;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.Statement;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import javax.inject.Inject;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.common.WarningHandlingLevel.AS_ERROR;
+import static com.facebook.presto.spi.StandardErrorCode.NOT_FOUND;
+import static com.facebook.presto.spi.StandardErrorCode.NOT_SUPPORTED;
+import static com.facebook.presto.spi.StandardErrorCode.WARNING_AS_ERROR;
+import static com.facebook.presto.sql.SqlFormatter.formatSql;
+import static com.facebook.presto.sql.analyzer.ConstantExpressionVerifier.verifyExpressionIsConstant;
+import static com.facebook.presto.sql.analyzer.SemanticErrorCode.INVALID_PARAMETER_USAGE;
+import static com.facebook.presto.sql.analyzer.utils.ParameterExtractor.getParameterCount;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+
+public class BuiltInQueryPreparer
+        implements QueryPreparer
+{
+    private final SqlParser sqlParser;
+
+    @Inject
+    public BuiltInQueryPreparer(SqlParser sqlParser)
+    {
+        this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
+    }
+
+    public BuiltInPreparedQuery prepareQuery(AnalyzerOptions analyzerOptions, String query, Map<String, String> preparedStatements, WarningCollector warningCollector)
+            throws ParsingException, PrestoException, SemanticException
+    {
+        Statement wrappedStatement = sqlParser.createStatement(query, analyzerOptions.getParsingOptions());
+        if (warningCollector.hasWarnings() && analyzerOptions.getWarningHandlingLevel() == AS_ERROR) {
+            throw new PrestoException(WARNING_AS_ERROR, format("Warning handling level set to AS_ERROR. Warnings: %n %s",
+                    warningCollector.getWarnings().stream()
+                            .map(PrestoWarning::getMessage)
+                            .collect(joining(System.lineSeparator()))));
+        }
+        return prepareQuery(analyzerOptions, wrappedStatement, preparedStatements);
+    }
+
+    public BuiltInPreparedQuery prepareQuery(AnalyzerOptions analyzerOptions, Statement wrappedStatement, Map<String, String> preparedStatements)
+            throws ParsingException, PrestoException, SemanticException
+    {
+        Statement statement = wrappedStatement;
+        Optional<String> prepareSql = Optional.empty();
+        if (statement instanceof Execute) {
+            String preparedStatementName = ((Execute) statement).getName().getValue();
+            prepareSql = Optional.ofNullable(preparedStatements.get(preparedStatementName));
```

Comment:
You definitely do not need this optional here. Checking for null is simpler., clearer, and mich easier to follow.

Reply:
Optional usage is preferred in the prestocode base instead of null checking. Also I don't want to touch prepareSQL signature in this change, its used at other places.